### ISSUE TITLE
Lambda adapter for callback with context

### ISSUE THREAD

**Issue by ysc3839** [microsoft/wil#143]
There's lots of APIs have callback with a context. For example `EnumWindows`, it has a `lParam` to be passed to the callback.

However, if we want to use C++ lambda with such function, we can't capture vars directly.
Instead, we have to do this:
```
int something_to_capture0 = 42;
int something_to_capture1 = 43;
int something_to_capture2 = 44;

struct {
	int* capture0;
	int* capture1;
	int* capture2;
} capture = { &something_to_capture0, &something_to_capture1, &something_to_capture2 };

EnumWindows([](HWND hWnd, LPARAM lParam) -> BOOL {
	auto _capture = reinterpret_cast<decltype(&capture)>(lParam);
	// Use captured vars here...
	return TRUE;
}, reinterpret_cast<LPARAM>(&capture));
```

So is there a "lambda adapter" class, which makes things like this?
```
int something_to_capture0 = 42;
int something_to_capture1 = 43;
int something_to_capture2 = 44;

wil::lambda_adapter<WNDENUMPROC> adapter{ [&](HWND hWnd) -> BOOL {
	// Use captured vars here...
	return TRUE;
} };
EnumWindows(adapter.func_ptr(), adapter.context());
```

**Comment by sylveon** [comment]
I like the idea and wonder if you could use it for member functions as well

**Comment by ChrisGuzak** [comment]
@oldnewthing I recall you did something like this. can you share what you learned with this thread?

**Comment by oldnewthing** [comment]
Writing a lambda adapter is hard. The above proposal has the problem that the lambda must be copied into the adapter, which means your captures must be copyable. But if you want to take them by reference, then you have the problem of taking a reference to a temporary. I have a lambda adapter that does a lot of template craziness to let you write
```cpp
template<typename TLambda>
auto EnumChildWindowsLambda(HWND hwnd, TLambda lambda)
{
  return EnumChildWindows(hwnd,
      LAMBDA_ADAPTER_ARGS(WNDENUMPROC, LPARAM, lambda));
}
```

but it does require to write out all those little helpers. It works, but I'm not super-proud of it.

I do like the above proposal. Will noodle on it.

**Comment by sylveon** [comment]
I guess it might be possible to use perfect forwarding so that

```cpp
wil::lambda_adapter<WNDENUMPROC> adapter{ [&](HWND hWnd) -> BOOL {
	// Use captured vars here...
	return TRUE;
} };
```

moves the lambda into the adapter, rather than copy.

**Comment by dunhor** [comment]
First to note that you _can_ use lambda capture, just not in the most ideal way. E.g.:
```c++
int count = 0;
auto proc = [&](HWND hwnd) -> BOOL
{
    std::printf("%3d: %p\n", count++, hwnd);
    return true;
};

::EnumWindows([](HWND hwnd, LPARAM ctxt)
{
    return (*reinterpret_cast<decltype(proc)*>(ctxt))(hwnd);
}, reinterpret_cast<LPARAM>(&proc));
```
Not ideal, but it actually requires minimal additional effort to the alternative (writing the forwarding lambda is tedious, but not _that_ bad). Something like the proposed has a few additional design difficulties. E.g.
* How is the 'context' argument "discovered?" It could be an `LPARAM` or it could be `void*` or a number of others. It could be the first argument or the last or in the middle (though that would be silly, but I wouldn't be surprised). As you note, an index is probably required
* Do you "remove" the context when forwarding the call? Doing so dramatically increases the complexity, especially without using `std::tuple`.

Here's one example of how this might be done without removing the context value: https://godbolt.org/z/7MKxG9. The use of `std::tuple`/`std::get` here is just convenience, but should be replaced to avoid the STL dependency. The `make_` helper is to avoid a dependency on `wistd::function` since that introduces more constraints, like the lambda must by copyable. As-is, it'll capture the lambda by reference (instead of copying), though that's easy to change if capturing by reference is seen as too easy to mess up - personally I'm fond of capturing by reference; you have to go to great lengths to get the holder object to outlive the lambda. I also haven't fully tested the forwarding of function arguments, but that should be fine.

As noted, this example also does not remove the context argument - mostly because anything I'd be able to come up with would be heavily reliant on `std::tuple`. If the STL dependency was deemed as worth it, it probably wouldn't be _too_ bad to formulate an `index_sequence` with the desired index removed, and since the number of arugments is relatively bounded, the impact on compile time is likely negligible (complex operations on `integer_sequence`s can take time...). I'm not entirely sure how one might correctly forward arguments, but that's probably not a big deal since most everything should be POD.

I'd also probably agree that having N hand-written wrapping functions seems a little out of place for WIL. That seems like something much better suited to automated generation, or possibly modifying the headers themselves with `#if __cplusplus`, etc.

**Comment by oldnewthing** [comment]
Yeah, autodetecting the function pointer and context parameter were the ugly parts of my lambda adapter.

**Comment by ysc3839** [comment]
@dunhor Thanks for telling me the "not ideal" way, which is a good workaround!

> Do you "remove" the context when forwarding the call?

I think the answer is NO. Since [the timer id is safe to hold a pointer](https://devblogs.microsoft.com/oldnewthing/20191009-00/?p=102974). We use the timer id as context, and then we may pass it to `KillTimer`. So the context param must not be removed.

**Comment by dunhor** [comment]
In this case, the context is being "hijacked" to hold the lambda pointer (or equivalent), so it's not very useful to virtually all consuming callbacks. What you might previously be passing as context would be something that I would expect to be captured in the lambda capture (or if you don't otherwise need the lambda capture, this feature is useless to you). And if what you want is "just some pointer", you could accomplish the same thing through a local variable that's captured by pointer/reference.

**Comment by ysc3839** [comment]
There's cases that the callback function references itself, for example `RemoveWindowSubclass`.
Is it possible to use lambda in such cases?

**Comment by oldnewthing** [comment]
You won't even be able to get that far because the lambda adapter, like your `capture` object, is going to be destructed after the call to `SetWindowSubclass` returns, leaving the callback to use a pointer to a destroyed object. The lambda adapter is useful only for synchronous callbacks, and those have no need for their own address.

**Comment by ysc3839** [comment]
@oldnewthing But what if I use it with a modal dialog?
For example:
```
auto adapter = ...
DialogBoxParamW(..., DialogFunc, (LPARAM)&adapter);

INT_PTR DialogFunc(...)
{
	if (Msg == WM_INITDIALOG)
	{
		auto adapter = (decltype(adapter)*)(lParam);
		SetWindowSubclass(..., adapter.function(), ..., adapter.context())
	}
}
```

Should I prevent using lambda in such case?

**Comment by oldnewthing** [comment]
If you can guarantee that the adapter doesn't destruct until all callbacks have returned, then great. But that's not the common scenario for the adapter.
